### PR TITLE
feat: cache analysed bytecodes

### PR DESCRIPTION
**Pull request by dinhani-cw**
We must test to see if this really improves performance before merging it.

### CODE REVIEW

**Review by renancloudwalk — DISMISSED** [review]

File: src/eth/evm/revm.rs [hunk 1]
```diff
@@ -126,16 +152,30 @@ struct RevmDatabaseSession {
 
     /// Changes made to the storage during the execution of the transaction.
     storage_changes: ExecutionChanges,
+
+    /// All analysed bytecodes that are cached.
+    all_analysed_bytecodes: Arc<RwLock<HashMap<BytecodeHash, RevmBytecode>>>,
+
+    /// New analysed bytecodes crated during this transaction execution.
+    new_analysed_bytecodes: Vec<(BytecodeHash, RevmBytecode)>,
```
Comment (renancloudwalk) [inline src/eth/evm/revm.rs:160]: shouldnt this be mut? I'm wondering how this is allowed :D

**Review by mayconamaroCW — DISMISSED** [review]

**Comment by carneiro-cw** [comment]
e2e-contracts:
  main -> 1m44s
  analyze_bytecode -> 1m40s
e2e-flamegraph (with the test data provided by @renancloudwalk):
 main -> flamegraph-main
  analyze_bytecode -> flamegraph-analyze

File: src/eth/evm/revm.rs [hunk 2]
```diff
@@ -102,7 +116,19 @@ impl Evm for Revm {
...
             Ok(result) => {
                 let session = evm.take_db();
-                Ok(parse_revm_execution(result, session.input, session.storage_changes)?)
+
+                // cache new bytecodes generated during this transaction
+                let has_new_bytecodes = session.has_new_bytecodes();
+                let result = parse_revm_execution(result, session.input, session.storage_changes)?;
+                if result.is_success() && has_new_bytecodes {
+                    let mut bytecodes_write_lock = self.bytecodes.write().unwrap();
+                    for (bytecode_hash, bytecode) in session.new_analysed_bytecodes {
```
Comment (carneiro-cw) [inline src/eth/evm/revm.rs:125]: You can use `HashMap::reserve` to allocate `session.new_analysed_bytecodes.len()` elements before beggining this loop.

**Comment by renancloudwalk** [comment]
```
p25_in_seconds                   | 0.009467
p50_in_seconds                   | 0.010397
p90_in_seconds                   | 0.22874440000000001
p99_in_seconds                   | 0.6150817199999953
p999_in_seconds                  | 1.9561200119999942
max_time_diff_seconds            | 2.121008
min_time_diff_seconds            | 0.007721
avg_time_diff_seconds            | 0.05085763364993215739
count_not_null_time_diff_seconds | 3685
```

seems like there is no performance gain here.

**Pull request closed without merging**